auth/api/iam/openid4vp.go

```diff
@@ -513,45 +528,66 @@ func (r Wrapper) validatePresentationNonce(presentations []vc.VerifiablePresenta
 	return returnErr
 }
 
-func (r Wrapper) handleAccessTokenRequest(ctx context.Context, verifier did.DID, authorizationCode *string, redirectURI *string, clientId *string) (HandleTokenRequestResponseObject, error) {
-	// first check redirectURI
-	if redirectURI == nil {
+func (r Wrapper) handleAccessTokenRequest(ctx context.Context, verifier did.DID, request HandleTokenRequestFormdataRequestBody) (HandleTokenRequestResponseObject, error) {
+	// check if redirectURI is present
+	if request.RedirectUri == nil {
 		return nil, oauthError(oauth.InvalidRequest, "missing redirect_uri parameter")
 	}
-	callbackURI, err := url.Parse(*redirectURI)
+	// check if code is present
+	if request.Code == nil {
+		return nil, oauthError(oauth.InvalidRequest, "missing code parameter")
+	}
+	// check if client state is present
+	if request.State == nil {
+		return nil, oauthError(oauth.InvalidRequest, "missing state parameter")
+	}
+	// check if code_verifier is present
+	if request.CodeVerifier == nil {
+		return nil, oauthError(oauth.InvalidRequest, "missing code_verifier parameter")
+	}
+	// check if client_id is present
+	if request.ClientId == nil {
+		return nil, oauthError(oauth.InvalidRequest, "missing client_id parameter")
+	}
+	// get initial session using state parameter
+	var serverSession OAuthSession
+	err := r.oauthServerStateStore().Get(*request.State, &serverSession)
 	if err != nil {
-		return nil, oauthError(oauth.InvalidRequest, "invalid redirect_uri parameter")
+		return nil, oauthError(oauth.InvalidRequest, "invalid or expired state")
+	}
+	// check if the code_verifier is valid
+	serverSession.PKCEParams.Verifier = *request.CodeVerifier
+	if !validatePKCEParams(serverSession.PKCEParams) {
+		return nil, oauthError(oauth.InvalidRequest, "invalid code_verifier")
 	}
-
 	// check if the authorization code is valid
 	var oauthSession OAuthSession
-	err = r.oauthCodeStore().Get(*authorizationCode, &oauthSession)
+	err = r.oauthCodeStore().Get(*request.Code, &oauthSession)
 	if err != nil {
 		return nil, oauthError(oauth.InvalidRequest, "invalid authorization code")
 	}
-
 	// check if the redirectURI matches the one from the authorization request
-	if oauthSession.redirectURI() != nil && oauthSession.redirectURI().String() != *redirectURI {
-		return nil, withCallbackURI(oauthError(oauth.InvalidRequest, fmt.Sprintf("redirect_uri does not match: %s vs %s", oauthSession.RedirectURI, *redirectURI)), callbackURI)
+	if oauthSession.redirectURI() != nil && oauthSession.redirectURI().String() != *request.RedirectUri {
```

Comment:
`RedirectUri` is optional in the access token request. It must only match if present. (And if it is present, this should also fail when `oauthSession.redirectURI() == nil`)

Reply:
removed